modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,217 @@
+		return appResult
+	}
+
+	types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeTimeout, callbackData.ContractAddr, callbackData.GasLimit, nil)
+	return appResult
+}
+
+// OnChanCloseConfirm defers to the underlying application
+func (im IBCMiddleware) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit defers to the underlying application
+func (im IBCMiddleware) OnChanCloseInit(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanOpenAck defers to the underlying application
+func (im IBCMiddleware) OnChanOpenAck(
+	ctx sdk.Context,
+	portID,
+	channelID,
+	counterpartyChannelID,
+	counterpartyVersion string,
+) error {
+	return im.app.OnChanOpenAck(ctx, portID, channelID, counterpartyChannelID, counterpartyVersion)
+}
+
+// OnChanOpenConfirm defers to the underlying application
+func (im IBCMiddleware) OnChanOpenConfirm(ctx sdk.Context, portID, channelID string) error {
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanOpenInit defers to the underlying application
+func (im IBCMiddleware) OnChanOpenInit(
+	ctx sdk.Context,
+	channelOrdering channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	channelCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	version string,
+) (string, error) {
+	return im.app.OnChanOpenInit(ctx, channelOrdering, connectionHops, portID, channelID, channelCap, counterparty, version)
+}
+
+// OnChanOpenTry defers to the underlying application
+func (im IBCMiddleware) OnChanOpenTry(
+	ctx sdk.Context,
+	channelOrdering channeltypes.Order,
+	connectionHops []string, portID,
+	channelID string,
+	channelCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (string, error) {
+	return im.app.OnChanOpenTry(ctx, channelOrdering, connectionHops, portID, channelID, channelCap, counterparty, counterpartyVersion)
+}
+
+// OnRecvPacket defers to the underlying application
+func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) ibcexported.Acknowledgement {
+	appAck := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	appAckResult, ok := appAck.(channeltypes.Acknowledgement)
+	if !ok {
+		return appAck
+	}
+
+	callbackData, err := types.GetCallbackData(im.app, packet, ctx.GasMeter().GasRemaining())
+	if err != nil {
+		types.EmitDestinationCallbackEvent(ctx, packet, types.CallbackTypeTimeout, callbackData.ContractAddr, callbackData.GasLimit, err)
+		return appAck
+	}
+
+	err = im.contractKeeper.IBCReceivePacketCallback(ctx, packet, callbackData.CustomMsg, appAckResult, relayer, callbackData.ContractAddr, callbackData.GasLimit)
+	if err != nil {
+		types.EmitDestinationCallbackEvent(ctx, packet, types.CallbackTypeTimeout, callbackData.ContractAddr, callbackData.GasLimit, err)
```

Comment:
looks like emission isn't happening here if err is nil. Think this can just be pulled out of the if statement as well